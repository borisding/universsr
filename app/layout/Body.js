import React from 'react';
import PropTypes from 'prop-types';
import { renderRoutes } from 'react-router-config';
import { ErrorBoundary } from '../common/components';
import './Body.module.scss';

export default function Body({ route }) {
  return (
    <div styleName="content">
      <ErrorBoundary>{renderRoutes(route.routes)}</ErrorBoundary>
    </div>
  );
}

Body.propTypes = {
  route: PropTypes.object.isRequired
};
