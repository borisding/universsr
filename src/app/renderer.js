import React from 'react';
import DocumentTitle from 'react-document-title';
import serialize from 'serialize-javascript';
import flushChunks from 'webpack-flush-chunks';
import { renderToString } from 'react-dom/server';
import { flushChunkNames } from 'react-universal-component/server';
import { matchRoutes, renderRoutes } from 'react-router-config';
import { StaticRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import storeFactory from './store';
import routes from './routes';

// preload data for matched route
function prefetchBranchData(store, url) {
  try {
    const branch = matchRoutes(routes, url);
    const promises = branch.map(({ route, match }) => {
      const { loadData } = route;
      const { dispatch } = store;

      if (match && match.isExact && loadData) {
        return Array.isArray(loadData)
          ? Promise.all(loadData.map(action => dispatch(action(match))))
          : dispatch(loadData(match));
      }

      return Promise.resolve(null);
    });

    return Promise.all(promises);
  } catch (err) {
    throw err;
  }
}

// export default server renderer and receiving stats
// also, should allow it to be mounted as middleware for production usage
export default function serverRenderer({ clientStats }) {
  return async (req, res, next) => {
    try {
      const context = {};
      const store = storeFactory();

      await prefetchBranchData(store, req.url);

      const appString = renderToString(
        <Provider store={store}>
          <StaticRouter location={req.url} context={context}>
            {renderRoutes(routes)}
          </StaticRouter>
        </Provider>
      );

      const pageTitle = DocumentTitle.rewind();
      const chunksOptions = { chunkNames: flushChunkNames() };
      const { scripts, stylesheets } = flushChunks(clientStats, chunksOptions);

      const preloadedState = serialize(store.getState(), { isJSON: true });
      const { statusCode = 200, redirectUrl } = context;
      const { nonce } = res.locals;

      if ([301, 302].includes(statusCode) && redirectUrl) {
        return res.redirect(statusCode, redirectUrl);
      }

      res.status(statusCode);
      res.render('index', {
        pageTitle,
        appString,
        preloadedState,
        stylesheets,
        scripts,
        nonce
      });
    } catch (err) {
      next(new Error(err));
    }
  };
}
