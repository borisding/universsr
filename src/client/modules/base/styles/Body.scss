@import '~@styles/variables';

.content {
  flex: 1 100%;
  padding: $base-padding * 4 $base-padding;
}
