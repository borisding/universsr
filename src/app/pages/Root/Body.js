import React from 'react';
import PropTypes from 'prop-types';
import Alert from 'react-s-alert';
import { renderRoutes } from 'react-router-config';
import { ErrorBoundary } from '@common/components';
import './styles/Body.scss';

const Body = ({ route }) => (
  <div styleName="content">
    <ErrorBoundary>
      {renderRoutes(route.routes)}
      <Alert
        stack={{ limit: 1 }}
        position="bottom"
        effect="flip"
        timeout={5000}
      />
    </ErrorBoundary>
  </div>
);

Body.propTypes = {
  route: PropTypes.object.isRequired
};

export default Body;
