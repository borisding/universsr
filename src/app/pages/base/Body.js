import React from 'react';
import PropTypes from 'prop-types';
import Alert from 'react-s-alert';
import { renderRoutes } from 'react-router-config';
import { ErrorBoundary, Loader } from '@common/components';
import './styles/Body.scss';

const Body = ({ route, isClient, isFetching }) => {
  return (
    <div styleName="content">
      <ErrorBoundary>
        {renderRoutes(route.routes)}
        {isClient && isFetching && <Loader />}
        <Alert
          stack={{ limit: 1 }}
          position="bottom"
          effect="flip"
          timeout={5000}
        />
      </ErrorBoundary>
    </div>
  );
};

Body.propTypes = {
  route: PropTypes.object.isRequired,
  isClient: PropTypes.bool.isRequired,
  isFetching: PropTypes.bool.isRequired
};

export default Body;
