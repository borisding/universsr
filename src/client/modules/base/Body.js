import React from 'react';
import PropTypes from 'prop-types';
import { renderRoutes } from 'react-router-config';
import Loader from '@common/components/Loader';
import './styles/Body.scss';

const Body = ({ route, isClient, isFetching }) => {
  return (
    <div styleName="content">
      {renderRoutes(route.routes)}
      {isClient && isFetching && <Loader />}
    </div>
  );
};

Body.propTypes = {
  route: PropTypes.object.isRequired,
  isClient: PropTypes.bool.isRequired,
  isFetching: PropTypes.bool.isRequired
};

export default Body;
