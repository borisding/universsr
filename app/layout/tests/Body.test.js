import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import Body from '../Body';

describe('Body', () => {
  it('renders correctly', () => {
    const props = { route: { routes: [] } };
    const tree = renderer
      .create(
        <MemoryRouter>
          <Body {...props} />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
