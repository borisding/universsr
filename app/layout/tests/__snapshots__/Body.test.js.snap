// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body renders correctly 1`] = `
<div
  className="content"
>
  <div
    className="s-alert-wrapper"
  />
</div>
`;
